### PR TITLE
Combine /problems and /filter

### PR DESCRIPTION
**Pull request by evancharlton**
This is a substantial change which overhauls the `/problems` and
`/filter` endpoints. The short version is that it removes `/filter` and
replaces it with client-side filtration of the data. (NOTE: `/filter`
still exists for backwards-compatibility.)

The advantages to this is that it happens client-side, so it's possible
to load the problems list once and then do the filtering offline. A side
advantage is that it removes a somewhat-redundant item in the
already-crowded navigation bar.

One disadvantage is that it happens client-side -- as the data set
grows, peformance might become an issue (on phones in particular). This
change includes performance monitoring of the critical path so we can
keep an eye on this and see if we need to revisit this approach. This
change was tested against the largest dataset (buldreinfo.com) and it
performed adequately, even with a 6x CPU throttle.

Finally, this change introduces some client-side caching of the
`/problems` data so that once the page is loaded, future visits to that
page should show up nearly instantly (with the data refreshed
asynchronously).

### CODE REVIEW

File: src/utils/grades.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
+/** Ranked from easiest to hardest */
+
+import { useCallback, useMemo } from "react";
+import { useMeta } from "../components/common/meta";
+
+export const SORTED_GRADES = {
```
Comment (evancharlton) [inline src/utils/grades.ts:6]: @jossi87 - I didn't know whether the grades that come out of `/meta` are sorted or not. If they're sorted into a consistent order (easiest -> hardest or hardest -> easiest) then I can remove all of this boilerplate and generate this mapping at runtime.

Is the client able to rely on this being a stable sort?

**Comment by evancharlton** [comment]
@jossi87 - I'll be very curious to hear what you think about this change. I'm also thinking of adding a separate render other than `<TableOfContents />`, being a map view as an alternate to the list view (kind of like how on Finn, Too Good To Go, etc you can view results in either way)

What do you think? There is a loss of functionality is that you can't filter by "has media" anymore - but I don't know whether that's important.

File: src/utils/grades.ts (hunk 1 above)
Comment (jossi87) [reply src/utils/grades.ts:6]: The grades from /meta is always ordered hardest -> easiest (consistent on every type: rock climbing, bouldering and ice climbing). You don't need the SORTED_GRADES :)

Comment (evancharlton) [reply src/utils/grades.ts:6]: Perfect - that'll make this much cleaner. I'll tidy this up :)

**Comment by jossi87** [comment]
I think it looks awesome! Much better than the Filter today. In my opinion you should just remove everything connected to the old Filter. I can remove the ws endpoint when this is done. The filter on "has media" is not necessary at all. And I agree on your proposal on an extra render for map.
90% of users are using cell phones. Can you tweak the design a bit to fit mobile users better? The new filter button takes a lot of horizontal space, and the subheader is not wrapping below it (forcing the subheader to take up 5 lines in my screenshot below). I also think that n/a should be before "3-" on the climbing grades, not after "9c" (it's located first on all other sites with grades). But I guess this is fixed when you rely on grade-order from ws (just invert it).
Can you also try to use three Segment's instead of nested Segment's? The page today has one big white segment, and a white bordered segment for filter located inside the main segment. I wonder if it looks better if you separate it into three different segments (with gray background between them). One for the header with Filter and Download-buttons, one for the filter and one for the data (the About page and Area pages utilizes multiple Segments).
I'll leave this pull request open until you have read my comment and given your feedback :)
image

**Comment by evancharlton** [comment]
All great points! I took a crack at addessing them in the change that I just pushed; here's a summary:

>The filter on "has media" is not necessary at all.
Great; it'll stay removed.

>And I agree on your proposal on an extra render for map.
Great! Would you like that in this change or in a follow-up change? I think a follow-up change would be better -- potentially one which combines `/areas` and `/problems`?

>90% of users are using cell phones. Can you tweak the design a bit to fit mobile users better? The new filter button takes a lot of horizontal space, and the subheader is not wrapping below it (forcing the subheader to take up 5 lines in my screenshot below).
Huh, this seems to be an existing bug which baffles me. However, if either button is **multiple words**, then it flows correctly. The fix was to tweak the Filter button to "Filter <things>" haha

>I also think that n/a should be before "3-" on the climbing grades, not after "9c"
Makes sense. I have no opinion about this and defer to your expertise (my thinking was that "n/a" is likely to be more on the "hard" side of the scale than the "easy" side 😀 )

>Can you also try to use three Segment's instead of nested Segment's? The page today has one big white segment, and a white bordered segment for filter located inside the main segment.
I tried splitting it up into a bunch of Segments but it wound up looking a little goofy - especially when the filter wasn't visible. However, I do agree about the single infinite white canvas. I took a stab at breaking that up. The logic is now:

- If the filter is visible, the header + filter is one segment, and the results are a different one
- If the filter is not visible, it stays as one segment

I'm not married to this setup though, and will happily make it look however you'd like 😀

**Comment by jossi87** [comment]
N/a is often a project, but also often a boulder or route with no information.
A follow up sounds good, do ut the way you think makes most sense 👍
Great hack 😃
I'll merge this now, I like the new approach!

**Pull request merged**